### ISSUE TITLE
`authUser` hangs with Persistent backend

### ISSUE THREAD

**Issue by parsonsmatt** [agrafix/users#5]
I'm working on an app that uses the Users library with the Persistent backend, and it's hanging when calling the `authUser` function.

In [this change](https://github.com/parsonsmatt/QuickLift/commit/0a6a14da3254fc96d63f8d8d1d4f203fbd352b86#diff-b7281b16e2b4f3144eb619c1b4f8e2f1R56), I set it up to generate a fresh Persistent/Postgresql connection pool and pass that in to the `authUser` function. I setup a bunch of `Debug.Trace` stuff in the `Web.Users.Persistent` module and this is the output I get when POSTing to the endpoint:

```
*Main Api Authentication Config Models Parser Server Users Users.TH Web.Users.Persistent> quickLift 
[Debug#SQL] SELECT "id", "created_at", "username", "email", "password", "active", "more" FROM "login" WHERE ((("username"=?)) OR (("email"=?))) LIMIT 1; [PersistText "asdf@asdf",PersistText "asdf@asdf"]
Just (Entity {entityKey = LoginKey {unLoginKey = SqlBackendKey {unSqlBackendKey = 8}}, entityVal = Login {loginCreatedAt = 2015-12-28 18:50:39.320844 UTC, loginUsername = "asdf@asdf", loginEmail = "asdf@asdf", loginPassword = "$2b$08$HYihAWQsPPCCi5dgrWjLk.NgdMy9ytUuy7w/Q8ATGyrXfJah/5Dr.", loginActive = True, loginMore = "{\"_userName\":\"asdf\"}"}})
in the if branch y'all
adsf
in createToken
Outside runPersistent tok: 04b1be5a-539d-480e-bcd6-77a6b94eaffe
```

This is what my debugging traces look like:

[authUser](https://github.com/parsonsmatt/users/blob/master/users-persistent/src/Web/Users/Persistent.hs#L193-L206)

[createToken](https://github.com/parsonsmatt/users/blob/master/users-persistent/src/Web/Users/Persistent.hs#L252-L265)

It never prints out the stuff before or after the `runPersistent` call, which makes me think that the pool is not being used properly for Persistent. Additionally, when I make a fresh pool and pass that in, other requests can be made. If I pass the normal pool in, all DB requests hang.

Creating users works just fine, as does listing them.

**Comment by agrafix** [comment]
Are you compiling the application with multithreading flags?

**Comment by parsonsmatt** [comment]
I am, though I was running in GHCi previously. I'm compiling and executing the code now, and additionally getting this line:

```
thread blocked indefinitely in an STM transaction
```

When I turn off the threaded options, then I get the same hanging behavior.

**Comment by parsonsmatt** [comment]
I ran the debugger and it's definitely hanging right when the nested call to `runPersistent` is done. It seems like there was a similar issue [here](https://github.com/agrafix/funblog/issues/8), though the code is lost to history. I'm working on a patch that'll shuffle IO actions around in the code to avoid nesting the calls.

**Comment by agrafix** [comment]
(for reference, see #6 )

**Comment by agrafix** [comment]
Fixed in #6
